src/transformers/models/dac/modeling_dac.py

```diff
@@ -0,0 +1,653 @@
+            Projected latents (continuous representation of input before quantization).
+    """
+
+    loss: torch.FloatTensor = None
+    audio_values: torch.FloatTensor = None
+    quantized_representation: torch.FloatTensor = None
+    codebook_indices: torch.FloatTensor = None
+    projected_latents: torch.FloatTensor = None
+
+
+@dataclass
+class DacEncoderOutput(ModelOutput):
+    """
+    Args:
+        loss (`torch.Tensor`):
+            Loss from the encoder model, comprising the weighted combination of the commitment and codebook losses.
+        quantized_representation (`torch.Tensor` of shape `(batch_size, dimension, time_steps)`, *optional*):
+            Quantized continuous representation of input.
+        codebook_indices (`torch.Tensor` of shape `(batch_size, num_codebooks, time_steps)`, *optional*):
+            Codebook indices for each codebook (quantized discrete representation of input).
+        projected_latents (`torch.Tensor` of shape `(batch_size, num_codebooks * dimension, time_steps)`, *optional*):
+            Projected latents (continuous representation of input before quantization).
+    """
+
+    loss: torch.FloatTensor = None
+    quantized_representation: torch.FloatTensor = None
+    codebook_indices: torch.FloatTensor = None
+    projected_latents: torch.FloatTensor = None
+
+
+@dataclass
+class DacDecoderOutput(ModelOutput):
+    """
+    Args:
+        audio_values (`torch.Tensor` of shape `(batch_size, 1, input_length)`):
+            Decoded audio data.
+    """
+
+    audio_values: torch.FloatTensor = None
+
+
+class Snake1d(nn.Module):
+    """
+    A 1-dimensional Snake activation function module.
+    """
+
+    def __init__(self, hidden_dim):
+        super().__init__()
+        self.alpha = nn.Parameter(torch.ones(1, hidden_dim, 1))
+
+    def forward(self, hidden_states):
+        shape = hidden_states.shape
+        hidden_states = hidden_states.reshape(shape[0], shape[1], -1)
+        hidden_states = hidden_states + (self.alpha + 1e-9).reciprocal() * torch.sin(self.alpha * hidden_states).pow(2)
+        hidden_states = hidden_states.reshape(shape)
+        return hidden_states
+
+
+class DacVectorQuantize(nn.Module):
+    """
+    Implementation of VQ similar to Karpathy's repo (https://github.com/karpathy/deep-vector-quantization)
+
+    Additionally uses following tricks from improved VQGAN
+    (https://arxiv.org/pdf/2110.04627.pdf):
+        1. Factorized codes: Perform nearest neighbor lookup in low-dimensional space
+            for improved codebook usage
+        2. l2-normalized codes: Converts euclidean distance to cosine similarity which
+            improves training stability
+    """
+
+    def __init__(self, config: DacConfig):
+        super().__init__()
+
+        self.in_proj = weight_norm(nn.Conv1d(config.hidden_size, config.codebook_dim, kernel_size=1))
+        self.out_proj = weight_norm(nn.Conv1d(config.codebook_dim, config.hidden_size, kernel_size=1))
```

Comment:
If I'm not mistaken, we usually apply `weight_norm` once when converting !
See how it's done for SpeechT5:
1. create utility methods to apply / remove weight norm (cf [here](https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/speecht5/modeling_speecht5.py#L3312-L3326))
2. when converting the weights, [apply the weight norm](https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/speecht5/convert_hifigan.py#L30) before loading the state dict then [remove the weight norm ](https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/speecht5/convert_hifigan.py#L55)when conversion is done


cf @amyeroberts, I just want to double check that this is correct!

Reply:
mmh I see, that's how it's defined in the original codebase, so I wonder if we shouldn't just leave it as it is... would also like to hear @amyeroberts opinion!

Reply:
I` would stick with how it's been done in other models in transformers, i.e. as suggested by @ylacombe. Even if it's done a certain way in the original repo, it's better to have consistency and part of bringing a model into the library it making these conversions